### ISSUE TITLE
Re-work `dg-output` test cases

### ISSUE THREAD

**Issue by tschwinge** [Rust-GCC/gccrs#1690]
A lot of our execution test cases use `dg-output` for verification.  That's OK if really necessary, but generally somewhat non-standard and fragile (I suppose): does matching `\n` work when the testing host/target system uses a different concept of line endings?  At least the `\n` should be changed into `[\n\r]\+`?

I suppose a lot of these tests might be re-written into some `assert` style, so that the result of the actual execution test already tells between PASS and FAIL, without the additional `dg-output` step.

**Comment by CohenArthur** [comment]
Definitely, it would be great to have the `assert!()` builtin for this

**Comment by dkm** [comment]
There are similar dg- directives for that. Maybe not as easy to use as what you have in rustc.
See https://gcc.gnu.org/onlinedocs/gccint/Directives.html#Directives for reference. There is:

```
{ dg-error regexp [comment [{ target/xfail selector } [line] ]] }

    This DejaGnu directive appears on a source line that is expected to get an error message, or else specifies the source line associated with the message. If there is no message for that line or if the text of that message is not matched by regexp then the check fails and comment is included in the FAIL message. The check does not look for the string ‘error’ unless it is part of regexp.

```
line can be relative using `+1` or `-2` so similar to your `~^^`.

**Comment by dkm** [comment]
yes, if a `dg-error` is not matched, the testcase is marked as FAIL (and you are notified of which exact test has been failing of course)

**Comment by bjorn3** [comment]
I mean does it check if there are no errors without a corresponding dg-error. For example if there is only a dg-error for the foo error, but the compiler emits both the foo and bar error, does the test fail?

This is not as important as failing when an expected error is not emitted, but it does help prevent unexpected regressions. Especially with respect to diagnostic quality.
